runtime-modules/project-token/src/types.rs

```diff
@@ -1,44 +1,89 @@
+/// either through purchase or during initial issuance
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub enum VestingSource {
+    InitialIssuance,
+    Sale(TokenSaleId),
+}
+
+/// Represent's account's split staking status
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub struct StakingStatus<Balance> {
+    // Id of the split
+    // TODO: SplitId
+    split_id: u32,
+
+    // The amount staked for the split
+    amount: Balance,
+}
+
+impl<Balance: Copy> StakingStatus<Balance> {
+    pub(crate) fn locks<BlockNumber>(&self, _b: BlockNumber) -> Balance {
+        // TODO: Implement
+        self.amount
+    }
+}
 
 /// Info for the account
 #[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
-pub struct AccountData<Balance, ReserveBalance> {
-    /// Non-reserved part of the balance. There may still be restrictions
-    /// on this, but it is the total pool what may in principle be
-    /// transferred, reserved and used for tipping.
-    pub(crate) free_balance: Balance,
+pub struct AccountData<VestingSchedule, Balance, StakingStatus, JoyBalance> {
+    /// Map that represents account's vesting schedules indexed by source.
+    /// Account's total unvested (locked) balance at current block (b)
+    /// can be calculated by summing `v.locks()` of all
+    /// VestingSchedule (v) instances in the map.
+    pub(crate) vesting_schedules: BTreeMap<VestingSource, VestingSchedule>,
+
+    /// Represents total amount of tokens held by the account, including
+    /// unvested and staked tokens.
+    pub(crate) amount: Balance,
 
-    /// This balance is a 'reserve' balance that other subsystems use
-    /// in order to set aside tokens that are still 'owned' by the
-    /// account holder, but which are not usable in any case.
-    pub(crate) stacked_balance: Balance,
+    /// Account's current split staking status
+    pub(crate) split_staking_status: Option<StakingStatus>,
 
     /// Bloat bond (in 'JOY's) deposited into treasury upon creation of this
     /// account, returned when this account is removed
-    pub(crate) bloat_bond: ReserveBalance,
+    pub(crate) bloat_bond: JoyBalance,
 }
 
 /// Info for the token
 #[derive(Encode, Decode, Clone, PartialEq, Eq, Default, Debug)]
-pub struct TokenData<Balance, Hash, BlockNumber> {
-    /// Current token supply
-    pub(crate) supply: Balance,
+pub struct TokenData<Balance, Hash, BlockNumber, TokenSale> {
+    /// Current token's total supply (tokens_issued - tokens_burned)
+    pub(crate) total_supply: Balance,
 
-    /// Initial offering state
-    pub(crate) offering_state: OfferingState,
+    /// Total number of tokens issued
+    pub(crate) tokens_issued: Balance,
+
+    // TODO: Limit number of sales per token?
+    /// Number of sales initialized, also serves as unique identifier
+    /// of the current sale (`last_sale`) if any.
+    pub(crate) sales_initialized: TokenSaleId,
```

Comment:
> Why does this live outside `TokenSale`? is there a case where we want to know the id of a past non-existing sale? if so why?

It's basically `next_sale_id`, it could also live in `TokenSale` as `id` (in which case `next_sale_id = last_sale.map_or(0, |s| s.id + 1)`) or be a separate global storage value, as you mentioned.

To compare the alternative approaches with the current one:
- If it lives in `TokenSale`, you need to perform this `let next_sale_id = token_data.last_sale.map_or(0, |s| s.id + 1)` which is slightly more complex than `let next_sale_id = token_data.sales_initialized + 1`, the storage write/read cost is pretty much the same though
- if `next_sale_id` lives as a separate storage value, there is some storage space saved (one `u32` value per each token), but additional storage read operation needs to be performed during sale initialization (in order to retrieve the `next_sale_id`)

Reply:
ah ok, lets rename to `next_sale_id`.

Reply:
Done in https://github.com/Joystream/joystream/pull/3618/commits/805aa09ac8dba3c3427738a06866668006415aea